lib/internal/modules/run_main.js

```diff
@@ -162,7 +157,34 @@ function runEntryPointWithESMLoader(callback) {
 function executeUserEntryPoint(main = process.argv[1]) {
   const resolvedMain = resolveMainPath(main);
   const useESMLoader = shouldUseESMLoader(resolvedMain);
-  if (useESMLoader) {
+
+  // We might retry as ESM if the CommonJS loader throws because the file is ESM, so
+  // try the CommonJS loader first unless we know it's supposed to be parsed as ESM.
```

Comment:
This is a little bit hard to understand. Can you simplify this a little bit to avoid confusion?